### PR TITLE
17 & 18: push code to repository with instructions on how to set up the development environment & Create shareable Docker image of iNatAtor

### PR DESCRIPTION
**Pull request by oz03-hub**
Providing starter code set up and Docker set up

Related to tasks 17 & 18:
- Provided start code for back-end FastAPI, in src/backend
- Provided starter code for frontend React, src/frontend
- Code added makes a simple request from frontend to backend
- Removed outdated cdstemplate package
- Removed sinr directory because Sergei is planning to add it in the backend directory @SergePogorelov 
Potential issue, if you want to re-add it as a package, we need to reconfigure pyproject.toml to do so

- Added Dockerfiles to backend and frontend as they are separate services
- Port numbers are still same, containers will run as they would locally
- Added docker-compose.yml to run and build container together
- Changed pyproject.toml to include dependencies we need
- Project now uses poetry as a dependency management tool to resolve package version conflicts
- README.md updated to include necessary instructions

You don't have to use docker during development, but it would solve 'works on my machine' problems

### CODE REVIEW

**Comment by oz03-hub** [comment]
Also linked in the ticket: https://hub.docker.com/repository/docker/oyilmazel/ds4cg-inaturalist/general
Images are hosted with instructions on how to run the application

**Comment by SergePogorelov** [comment]
image
There seems to be a small typo here?

**Comment by SergePogorelov** [comment]
@oz03-hub, @cheerstopaula 
should we also remove all template files and readme instructions which is not related to our app?

**Comment by oz03-hub** [comment]
@SergePogorelov I don't have a reason to hold them I just like looking at instructions sometimes because it explains things in the repo later in the sprint we can erase them

**Comment by oz03-hub** [comment]
I learned that there is no longer need for version line in docker-compose, erased that in last commit

**Review by cheerstopaula — APPROVED** [review]

File: pyproject.toml [hunk 1]
```diff
@@ -18,16 +15,6 @@ torch = "^1.9.0"
 fastapi = "^0.68.0"
 uvicorn = {extras = ["standard"], version = "^0.14.0"}
 
-[tool.poetry.extras]
-test = ["pytest"]
-dev = [
-    "ruff",
```
Comment (ginic) [inline pyproject.toml:24]: It's probably a good idea to keep using development tools like pytest and ruff.

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks @oz03-hub! The Dockerfiles and docker compose configuration look great. I also think you did a good job structuring the repository and README to make it clear to developers where the frontend/backend code is and how get started using it. 

I do have some concerns about the python environment and dependencies. I've summarized them here, but please read the in-line comments in more detail to better understand the concerns. 
- The versions of some of the python libraries being used are quite old, several from 2021. By using older versions, we are missing out on bug and security fixes and likely to run into more environment issues that using later versions. Please update these to later versions and test the environment again with the docker containers to make sure everything still works. 
- The poetry installation doesn't work on new Macs, which is a problem for your fellow developers. Poetry is a nice tool and if the team would like to continue using it, that is fine with me. However, I do think that all team members need to be able to use it effectively for their local development. I've included some links on how to get it working. Poetry is a newer tool, and so will have more maintenance and learning overhead than more established and widely used tools. Adding new tools like this should be carefully considered and discussed with the team. If you decide not to use poetry, you can use conda or even just pip to share requirements/environments files in the repository. You will also need to update the Docker files according to whatever you use and re-test. 
- Many of the unit testing and linting tools have been removed from the repository. These are important parts of best practices going forward, so I think it's better to keep them if possible, unless getting them to work is creating a lot of extra work and frustration. Whether we use poetry, conda or pip, we should plan on there being multiple environment set ups, one for development/testing and one for deployment. For example, pip supports this as https://stackoverflow.com/questions/17803829/how-to-customize-a-requirements-txt-for-multiple-environments#20720019. Relatedly, I would like the ruff checks on Github to be passing before merging.

File: README.md [hunk 2]
```diff
@@ -1,134 +1,55 @@
...
 
 # Getting Started
-## Installation for development
+## :hatched_chick: Installation for local development
 Make sure you update your local branch to the latest.
 
 ```
 $ conda create -n inatator python=3.9
 $ conda activate inatator
 ```
```
[the remark this answers is not in the document]
Comment (oz03-hub) [reply README.md:16]: using only conda and pip now

File: README.md [hunk 3]
```diff
@@ -1,134 +1,55 @@
...
+1. Check if you have `poetry`, this is needed because poetry is a tool that finds all compatible library versions while downloading them, otherwise we would die
+  - To install poetry go to https://python-poetry.org/docs/
+  - It will need you to install pipx, pip on steroids, follow the official pipx instructions to install
+  - You don't have to run the command with --global option, the last command
+  - Steps will differ slightly based on your OS
+  - After pipx is installed you can install poetry from their instructions, just one line
+2. Check poetry by just typing `poetry` into terminal, it should recognize and list options etc
+
+3. Now at the project root you can run `poetry install` which will install all packages required for now
+  - To double check everything is working as intended, navigate to `src/sinr/models.py` write a function that prints something. Now, navigate to `src/backend/app` and create a temporary python file that you will delete later, in the file enter `from sinr.models import *`, then call the function you just created from that file, make sure it prints correctly and does not give errors.
```
[the remark this answers is not in the document]
Comment (oz03-hub) [reply README.md:32]: sinr as a package was removed, instead will be included in backend, sergei is actively working on its integration, in future sprints we can work on tests

File: README.md [hunk 4]
```diff
@@ -1,134 +1,55 @@
...
+### Backend and SINR Dependency Installation
+1. Check if you have `poetry`, this is needed because poetry is a tool that finds all compatible library versions while downloading them, otherwise we would die
+  - To install poetry go to https://python-poetry.org/docs/
+  - It will need you to install pipx, pip on steroids, follow the official pipx instructions to install
+  - You don't have to run the command with --global option, the last command
+  - Steps will differ slightly based on your OS
+  - After pipx is installed you can install poetry from their instructions, just one line
+2. Check poetry by just typing `poetry` into terminal, it should recognize and list options etc
+
+3. Now at the project root you can run `poetry install` which will install all packages required for now
```
[the remark this answers is not in the document]
Comment (oz03-hub) [reply README.md:31]: thanks, there will be two requirement files, and stopped using poetry

[the remark this answers is not in the document]
Comment (oz03-hub) [reply README.md:31]: miniconda3 image works, the build image comes to a whopping 10gb, I am not sure why it's so big

File: src/backend/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ENV PYTHONDONTWRITEBYTECODE 1
+ENV PYTHONUNBUFFERED 1
+
+WORKDIR /app
+
+RUN conda create -n inatator python=3.10
+
+COPY . .
+
+RUN pip install -r requirements.txt && pip install -r requirements-dev.txt
```
Comment (ginic) [inline src/backend/Dockerfile:12]: Make sure that the dependencies get installed to the conda environment you created. Since we're only using the Docker containers for deployment, you don't need to install the dev dependencies here. 

```suggestion
RUN conda run -n inatator \
    pip install -r requirements.txt
```

**Review by ginic — CHANGES_REQUESTED** [review]
One minor fix in the docker container! After that's done, it looks good to me!

File: .github/dependabot.yml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# To get started with Dependabot version updates, you'll need to specify which
```
[the remark this answers is not in the document]
Comment (oz03-hub) [reply .github/dependabot.yml]: removed 👍🏼

File: src/backend/Dockerfile (hunk 5 above)
Comment (oz03-hub) [reply src/backend/Dockerfile]: Thanks for the catch, I added your suggestion and was able to successfully build

**Review by ginic — APPROVED** [review]
Great work! This all works for me. Thank you for remembering to update the CHANGELOG and README too!

File: README.md [hunk 7]
```diff
@@ -1,258 +1,45 @@
...
-
-## A Note on Notebooks
-We have also included an example Jupyter notebook
 
-Jupyter notebooks are useful tools for exploratory data analysis, prototyping baseline models and creating visualizations. However, they are _not_ an acceptable way to hand-off code for others to reproduce. Have you ever tried to run someone else's notebook, only to find out a cell was deleted, and you have no idea what it was supposed to do?
+3. At the root of the project `pip install -r src/backend/requirements.txt` and `pip install -r src/backend/requirements-dev.txt`.
 
-[Don't put data science notebooks into production](https://martinfowler.com/articles/productize-data-sci-notebooks.html), they are [hard to test, version, parametrize and keep track of state](https://www.reddit.com/r/datascience/comments/ezh50g/jupyter_notebooks_in_productionno_just_no/).
+## :penguin: Run the Application
+Open two terminals. You will need to run the server first, then application frontend.
```
Comment (cheerstopaula) [inline README.md:29]: Maybe mention that conda environment needs to be activated for this steps.

File: README.md [hunk 8]
```diff
@@ -1,258 +1,45 @@
...
+1. ***Run the backend (server):*** To start the server, navigate to project root, run `uvicorn src.backend.app.main:app --reload`, to make sure it is working go to `http://localhost:8000/hello/` in your browser.
```
Comment (cheerstopaula) [inline README.md:30]: What does it mean to 'it is working'? Someone might see the page and not be able to tell if that is what they are supposed to be seeing. Add description or image of what 'working' looks like.

File: README.md [hunk 9]
```diff
@@ -1,258 +1,45 @@
...
+2. ***Run the frontend (application):*** To start the application, navigate to `src/frontend/` you see there are js files, you are at the right spot. Now run `npm start`, it will start running the app and should automatically open to page in your browser.
 
-There _are_ [companies that use notebooks in production architecture](https://blog.goodaudience.com/inside-netflixs-notebook-driven-architecture-aedded32145e), but they have entire Devops organizations to help configure deployment and _still_ use workflow tools like [papermill](https://papermill.readthedocs.io/en/latest/) and Airflow to parametrize notebook dependencies. Unless you are willing to put in the effort to parametrize your notebooks in pipeline workflows, don't use them when stability and reproducibility matter.
+### Running applications with Docker
+1. Install Docker if you haven't already
```
Comment (cheerstopaula) [inline README.md:34]: Not enough to have it installed, it also needs to be running. Add to the instruction '.. and open the app' or something.

File: README.md [hunk 10]
```diff
@@ -1,258 +1,45 @@
...
+2. Navigate project root
+3. Run `docker compose up --build`, for the first build it may take a while
+4. Make sure the default application is running as expected by going to urls for react and server
```
Comment (cheerstopaula) [inline README.md:37]: This instruction is not entirely clear, could you rephrase or expand on what 'going to urls for react and server' mean?

**Review by SergePogorelov — APPROVED** [review]
good work!

**Comment by oz03-hub** [comment]
Merging, all changes approved, reviewed
